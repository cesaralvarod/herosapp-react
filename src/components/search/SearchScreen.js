import React, { useMemo } from "react";
import queryString from "query-string";
import { useLocation } from "react-router";
import { useForm } from "../../hooks/useForm";
import HeroCard from "../heroes/HeroCard";
import { getHerosByName } from "../../selectors/getHerosByName";

const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({ search: "" });

  const { search } = formValues;

  const heros = useMemo(() => getHerosByName(q), [q]);

  const handleSearch = (event) => {
    event.preventDefault();
    history.push(`?q=${search}`);
  };

  return (
    <div className="container mt-5 mb-5">
      <div className="row">
        <div className="col-5">
          <h4>Search</h4>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              name="search"
              placeholder="Find your hero"
              className="form-control"
              autoComplete="off"
              onChange={handleInputChange}
            />
            <button
              type="submit"
              className="btn mt-3 btn-block btn-outline-primary"
            >
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {q === "" && <div className="alert alert-info">Search a hero</div>}
          {q !== "" && heros.length === 0 && (
            <div className="alert alert-danger">No results of {q}</div>
          )}
          {heros.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
