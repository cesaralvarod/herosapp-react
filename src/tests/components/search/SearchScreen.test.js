import React from "react";
import { mount } from "enzyme";
import SearchScreen from "../../../components/search/SearchScreen";
import { MemoryRouter, Route } from "react-router-dom";

describe("Testing in SearchScreen", () => {
  test("Component should render correctly", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
  });

  test("Component should show batman", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find("h5.card-title").text().trim()).toBe("Batman");
  });

  test("Component should show an error if hero is not found", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=sometext"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find("div.alert.alert-danger").text().trim()).toBe(
      "No results of sometext"
    );
  });

  test("Component should call push of history", () => {
    const history = {
      push: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=sometext"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );

    const event = {
      target: {
        name: "search",
        value: "superman",
      },
    };

    wrapper.find("input").simulate("change", event);

    wrapper.find("form").simulate("submit", { preventDefault() {} });

    expect(history.push).toHaveBeenCalledWith(`?q=${event.target.value}`);
  });
});
