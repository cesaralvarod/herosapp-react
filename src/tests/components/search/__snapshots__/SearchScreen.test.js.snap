// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in SearchScreen Component should render correctly 1`] = `
<div
  className="container mt-5 mb-5"
>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
        />
        <button
          className="btn mt-3 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing in SearchScreen Component should show an error if hero is not found 1`] = `
<div
  className="container mt-5 mb-5"
>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
        />
        <button
          className="btn mt-3 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No results of 
        sometext
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing in SearchScreen Component should show batman 1`] = `
<div
  className="container mt-5 mb-5"
>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
        />
        <button
          className="btn mt-3 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3 mt-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters "
        >
          <div
            className="col-md-4 "
          >
            <img
              alt="Batman"
              className="card-img"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                More...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
